### ISSUE TITLE
plugin/transfer: New plugin for handling zone transfers

### ISSUE THREAD

**Issue by chrisohaver** [coredns/coredns#3121]
As discussed in #2585, the idea here is to implement zone transfers as a separate plugin.
Distilling what was discussed in #2585 ...

Syntax would look like ...
```
transfer [ZONES] {
    to HOSTS
}
```
 **ZONE** - defines a space delimited list of zones the plugin will handle.  If ZONES  is empty, all zones of the serverblock are handled.
**to HOSTS** - defines to which HOSTS transfers can be sent

The plugin...

* Defines interface `Transferer` that plugins must implement to enable zone transfers, comprising ...
   * `Records()` - returns all records
   * `SOA()` - returns SOA record
   * `Authoritative()` - Returns list of zones the plugin is authoritative for
* Is placed before other plugins in plugin.cfg. It would only handles transfer requests, falling through for other query types. 
   * Alternately it could be placed after, but then other plugins would need to fall through for transfer type queries.
* When processing a request, _transfer_ checks for plugins that implement `Transferer`.  Constructs zone transfer response from `Records()` and `SOA()` of first plugin that is `Authoritative()` for the requested zone.
* periodically checks `SOA()` for all plugins in the chain (with `Authoritative()` zones in ZONES), and sends notifies if SOA is changed.

Existing plugins (e.g. *kubernetes*) that already implement zone transfer should be updated to use the new interface, and docs updated.

See #2585 for prior discussion and context.

**Comment by chrisohaver** [comment]
@miekg, @ramnes, @JensErat

**Comment by chrisohaver** [comment]
If this hasn't been started, I can take a whack at it.

This is on my radar because I'd like to add zone transfer to _k8s_external_.

**Comment by ramnes** [comment]
Awesome. I personally did not write a single line of code on this yet. I'd be glad to give my humble review if needed, or any other sort of help.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] plugin/transfer: ..." ]
>As discussed in #2585, the idea here is to implement zone transfers as a separate plugin.
>Distilling what was discussed in #2585 ...
>
>Syntax would look like ...

This looks entirely reasonable.

It only covers *outgoing* transfers? 

Not sure about separating the notify hosts from the transfer to hosts.

**Comment by chrisohaver** [comment]
> It only covers *outgoing* transfers?

Yes, for now at least.  Perhaps could be expanded to include _incoming_ in the future, e.g. as an additional interface `TransferReceiver`.

> Not sure about separating the notify hosts from the transfer to hosts.

Makes sense.  I think the use case that you'd only want to send NOTIFYs to _some_ of the destinations is probably a corner case we don't need to handle. So, instead...

```
transfer [ZONE...] {
    to HOST...
    notify
}
```
`to` being a required option (unless we are ok with defaulting to `*`)
`notify` being optional

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>```
>transfer [ZONE...] {
>    to HOST...
>    notify
>}
>```
>`to` being a required option (unless we are ok with defaulting to `*`)
>`notify` being optional

notifies are super lightweight, I don't think anyone would care enough to make 
this a separate option.

**Comment by chrisohaver** [comment]
> I don't think anyone would care enough to make this a separate option.

OK, always send notifies = Less options = good.

**Comment by JensErat** [comment]
I haven't started work on this apart from the concept/idea so far. This was one of two findings blocking me from entirely moving my old DNS zones from bind to CoreDNS -- the other is dnssec for transfers (and this is the more important for me). Both are somewhat related though. I like that you want to start doing some progress -- feel free to get me into the loop, I might use the chance to get some hands on regarding better dnssec support for transfers. This is probably something that should happen _after_ the transfer plugin is ready, though (just something to keep in mind).

I just had much more important projects than migrating my private bind setup to CoreDNS -- my backlog starts clearing up though, so I might be ready to add up some effort soon. But feel free to claim the plugin -- I mostly want to have it, there is enough code to write. ;)

Your design doc looks good to me -- just one though:

> periodically checks SOA() for all plugins in the chain (with Authoritative() zones in ZONES), and sends notifies if SOA is changed.

Maybe we can rather implement this the other way round -- plugins that actually define an SOA and implement the transfer interface _notify_ the actual transfer/notify plugin about changed SOA (channel passed through or something like this?) instead the transfer/notify plugin polls. Seems cleaner, if we can somehow pass the channel in the caddy plugin chain model.

> notifies are super lightweight, I don't think anyone would care enough to make
this a separate option.

There is another issue though (at least I have it) -- the CoreDNS server is running behind a failover IP. This can be achieved by multiple IP addresses on the same interface or multiple interfaces -- in both cases, the default route might be not the one we need to send the notify from. It would be great if we'd be able to define the IP address used for sending the notify, resolving this on the routing/firewall side requires very quirky hacks.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>CoreDNS -- the other is dnssec for transfers (and this is the more important 
>for me). Both are somewhat related though. I like that you want to start doing 

DNSSEC for transfers isn't a thing in the DNS, what do you mean by this?

>> periodically checks SOA() for all plugins in the chain (with Authoritative() 
>> zones in ZONES), and sends notifies if SOA is changed.
>
>Maybe we can rather implement this the other way round -- plugins that actually 
>define an SOA and implement the transfer interface _notify_ the actual 
>transfer/notify plugin about changed SOA (channel passed through or something 
>like this?) instead the transfer/notify plugin polls. Seems cleaner, if we can 
>somehow pass the channel in the caddy plugin chain model.

this check only needs to happen when a AXFR request comes (with maybe some 
ratelimiting).

>be able to define the IP address used for sending the notify, resolving this on 
>the routing/firewall side requires very quirky hacks.

agreed, but this doesn't need to happen here and now; once this is in we can 
work on this; not sure miekg/dns supports this though.

**Comment by JensErat** [comment]
> DNSSEC for transfers isn't a thing in the DNS, what do you mean by this?

Currently, the DNSSEC plugin only signs data for usual "client-side requests" like an A or AAAA request, but not a zone transfer. If you want to sign a zone though, you also have to make sure the AXFR request is answered with signed zone data. This is currently not implemented by CoreDNS.

This is what we discussed around March: https://github.com/coredns/coredns/issues/2585#issuecomment-470835694

**Comment by miekg** [comment]
This is orthogonal from the transfer plugin. See my sign plugin PR for a
DNSSEC solution

**Comment by JensErat** [comment]
Is it? What about signing transfers for etcd (DynDNS-style applications, ...), or even Kubernetes (if that is to be considered a use case)? The `dnssec` plugin is there to "enable on-the-fly DNSSEC signing of served data" -- which would pretty well also fit ad-hoc signing of transfers.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>Is it? What about signing transfers for etcd (DynDNS-style applications, ...), or even Kubernetes (if that is to be considered a use case)? The `dnssec` plugin is there to "enable on-the-fly DNSSEC signing of served data" -- which would pretty well also fit ad-hoc signing of transfers.

do you mean tsig or something? DNSSEC has nothing to do with zone transfers

**Comment by chrisohaver** [comment]
FYI guys - I have a work in progress here #3223... PTAL, and chime in if you think its going in the wrong direction, or have any other comments/suggestions.

**Comment by mimuret** [comment]
if error happened when transfer implement plugin fetches records from data source.
How do you handle it?

**Comment by chrisohaver** [comment]
> if error happened when transfer implement plugin fetches records from data source.
How do you handle it?

It would return a SERVFAIL to the client requesting the transfer.

**Comment by mimuret** [comment]
I think It is possible when func Transfer before finished. (I should return error)
I want know how to handle error happened when reading data and sending to channel.
(ex. lost database connection)

**Comment by chrisohaver** [comment]
> I think It is possible when func Transfer before finished. (I should return error)
I want know how to handle error happened when reading data and sending to channel.
(ex. lost database connection)

As the transfer plugin is now: You could implement `Transferer.Transfer()` in such a way that all of the records are gathered before sending them to the channel.  If an error is encountered before gathering all the records (e.g. db conn lost), then return an error.  If all records are gathered successfully, then send the records to the channel.  

But doing it that way means you cannot gather the records from the source in a go routine. We could alter the transfer plugin to handle this by adding an error to the channel (making the channel a struct of []dns.RR and and error )... or do a separate error channel.  I don't know which way is more idiomatic.

**Comment by mimuret** [comment]
I think making the channel a struct is good.
Because Transfer.Out() part of dns library implement same method.

**Comment by valkolaci** [comment]
> [ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
> ``` transfer [ZONE...] { to HOST... notify } ``` `to` being a required option (unless we are ok with defaulting to `*`) `notify` being optional
> notifies are super lightweight, I don't think anyone would care enough to make this a separate option.

[ERROR] plugin/file: notify for zone "example.com." was not accepted by "193.194.195.196:53": rcode was "REFUSED"

Notifies are super lightweight - the error messages in the logs are not.

**Comment by miekg** [comment]
I've still haven't made the time to proper check all this code, mostly because life got in
the way (covid, moving house and the like). This is still on the todo - and I'm hoping we
can release 1.7.0 with these bits all included.

@stp-ip can you check out this branch and see if it works for you, esp, since you're using
'auto' which I personally are not using. Thanks!

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>Notifies are super lightweight - the error messages in the logs are not.

 ???

**Comment by stp-ip** [comment]
Will give it a go this week. Had it on my radar, but not enough time to take a full jump into it.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>Will give it a go this week. Had it on my radar, but not enough time to take a
>full jump into it.

I know that feeling :)

**Comment by stp-ip** [comment]
Just that I have the full picture of PRs and branches:
transfer plugin: https://github.com/coredns/coredns/pull/3223 [reviewed]
transfer with auto+file: https://github.com/coredns/coredns/pull/3732 [tested file and auto transfers including having legacy transfers configured, reviewed]
transfer with kubernetes: https://github.com/coredns/coredns/pull/3452 [tested, reviewed]
transfer with k8s_external: https://github.com/coredns/coredns/pull/3419 [reviewed]
transfer notifies: https://github.com/coredns/coredns/pull/3420
transfer notifies with auto: https://github.com/coredns/coredns/compare/master...chrisohaver:auto-file-notify

**Comment by chrisohaver** [comment]
> Just that I have the full picture of PRs and branches:

Yes. Thanks @stp-ip!

I also implemented transfer for  kubernetes and k8s_external ...
#3419 - transfer for _k8s_external_
#3452 - transfer for _kubernetes_

#3419 is the only change that actually adds new functionality, since previously _file/auto/kubernetes_ each had independently implemented transfer functionality. Adding transfer capability to _k8s_external_ was my primary driver for adding the transfer plugin (related: #2528).

**Comment by stp-ip** [comment]
Idea for introducing the full changes into CoreDNS:

With 1.7.0 transfer plugin + related plugin interfaces
With 1.7.x add notify functionality and deprecate transfer usage within plugins in docs
With 1.8.0 deprecate transfer usage within plugins with error/warning
With 1.9.0 remove transfer usage by plugins

Sidenote: Moving through the reviews and testing one by one ^^

**Comment by chrisohaver** [comment]
> With 1.7.x add notify functionality and deprecate transfer usage within plugins in docs

I'd like to get the notify functionality into 1.7.0 if possible.  I have working notify implementations for auto, file, k8s_external, and kubernetes in local branches.  I had not pushed them up as PRs because they are dependent on #3420 (and the respective transfer implementations), which have not merged yet.
